File: year-2021/day-02/solution.py
with open("input-01.txt") as fp:
    inp = [ i.split(' ') for i in fp.readlines() ]
    inp = [ (i[0], int(i[1])) for i in inp ]

    x = 0
    d = 0
    for (inst, dist) in inp:
        if inst == 'forward':
            x = x + dist
        elif inst == 'down':
            d = d + dist
        elif inst == 'up':
            d = d - dist
    print(x * d)

    x = 0
    d = 0
    a = 0
    for (inst, dist) in inp:
        if inst == 'forward':
            x = x + dist
            d = d + (dist * a)
        elif inst == 'down':
            a = a + dist
        elif inst == 'up':
            a = a - dist
    print(x * d)

File: year-2021/day-02/solution.rs
#[derive(Debug, PartialEq)]
pub enum Instruction {
    Forward(u64),
    Down(u64),
    Up(u64),
}

peg::parser! {
    grammar instruction_parser() for str {
        rule number() -> u64 = n:$(['0'..='9']+) {? n.parse().or(Err("u64")) }

        rule forward() -> Instruction = "forward " n:number() { Instruction::Forward(n) }
        rule down() -> Instruction = "down " n:number() { Instruction::Down(n) }
        rule up() -> Instruction = "up " n:number() { Instruction::Up(n) }

        pub rule line() -> Instruction = forward() / up() / down()
    }
}

fn main() {
    let instructions: Vec<_> = include_str!("input-01.txt").lines()
        .filter_map(|s| instruction_parser::line(s).ok())
        .collect();

    let mut x = 0u64;
    let mut d = 0u64;
    for instruction in &instructions {
        match instruction {
            Instruction::Forward(dist) => { x += dist; },
            Instruction::Down(dist) => { d += dist; },
            Instruction::Up(dist) => { d -= dist; },
        }
    }
    println!("Part 1: {}", x * d);

    let mut x = 0u64;
    let mut d = 0u64;
    let mut a = 0u64;
    for instruction in &instructions {
        match instruction {
            Instruction::Forward(dist) => {
                x += dist;
                d += dist * a;
            },
            Instruction::Down(dist) => { a += dist; },
            Instruction::Up(dist) => { a -= dist; },
        }
    }

    println!("Part 2: {}", x * d);
}